src/canopen_comm.cpp

```diff
@@ -0,0 +1,191 @@
+            throw -1;
+        }
+
+        struct ifreq ifr;
+
+        std::strcpy(ifr.ifr_name, ifname.c_str());
+        ioctl(_socket_handle, SIOCGIFINDEX, &ifr);
+
+        struct sockaddr_can addr;
+        
+        addr.can_family  = AF_CAN;
+        addr.can_ifindex = ifr.ifr_ifindex;
+
+        std::cout << ifname << " at index " << ifr.ifr_ifindex << std::endl;
+
+        struct can_filter rfilter[2];
+
+        rfilter[0].can_id   = _sdo_cob_id_offset + roboteq::canopen_comm::_roboteq_can_id;
+        rfilter[0].can_mask = CAN_SFF_MASK;
+        rfilter[1].can_id   = _sdo_response_cob_id_offset + roboteq::canopen_comm::_roboteq_can_id;
+        rfilter[1].can_mask = CAN_SFF_MASK;
+
+        setsockopt(_socket_handle, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter));
+
+        if(bind(_socket_handle, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+            printf("Error in socket bind");
+            throw -2;
+        }
+    
+    }
+
+    bool canopen_comm::sdo_download(send_runtime_command command, uint8_t subindex, uint32_t data) {
+
+        struct can_frame frame;
+
+        frame.can_id  = _sdo_cob_id_offset + roboteq::canopen_comm::_roboteq_can_id;
+        frame.can_dlc = 8;
+        frame.data[0] = (_sdo_command<<4) | (canopen_comm::_runtime_command_map.at(command).number_of_unused_bytes<<2);
+        frame.data[1] = canopen_comm::_runtime_command_map.at(command).canopen_index; 
+        frame.data[2] = canopen_comm::_runtime_command_map.at(command).canopen_index >> 8;
+        frame.data[3] = subindex;
+        frame.data[4] = data >> 0;
+        frame.data[5] = data >> 8;
+        frame.data[6] = data >> 16;
+        frame.data[7] = data >> 24;
+
+        write(roboteq::canopen_comm::_socket_handle, &frame, sizeof(struct can_frame));
+
+        struct can_frame response_frame = {};
+
+        uint32_t nbytes = read(roboteq::canopen_comm::_socket_handle, &response_frame, sizeof(struct can_frame));
+
+        std::cout <<  std::hex << response_frame.can_id << "\t" << static_cast<unsigned>(response_frame.can_dlc) << "\t" << static_cast<unsigned>(response_frame.data[0])
+                                << "\t" << static_cast<unsigned>(response_frame.data[1]) << "\t" << static_cast<unsigned>(response_frame.data[2]) << "\t" << static_cast<unsigned>(response_frame.data[3]) 
+                                << "\t" << static_cast<unsigned>(response_frame.data[4]) << "\t" << static_cast<unsigned>(response_frame.data[5]) << "\t" << static_cast<unsigned>(response_frame.data[6])
+                                << "\t" << static_cast<unsigned>(response_frame.data[7]) << std::endl;
+    
+        if (nbytes != sizeof(struct can_frame)) {
+                printf("Invalid CAN frame\n");
+                return 0;
+                }
+        else {
+                return (response_frame.data[4] | response_frame.data[5] << 8 | response_frame.data[6] << 16 | response_frame.data[7] << 24);
+                }
+    }
```

Comment:
sdo_download only returns a bool that indicates successful download or not. the response data frame should always be 0s i think